core/util.go

```diff
@@ -208,6 +208,32 @@ func GetBuildHost() (retID string) {
 	return
 }
 
+// IsAnyNilOrZero returns whether any of the supplied values are nil, or (if not)
+// if any of them it its type's zero-value. This is useful for validating that
+// all required fields on a proto message are present.
+func IsAnyNilOrZero(vals ...interface{}) bool {
+	for _, val := range vals {
+		switch v := val.(type) {
+		case nil:
+			return true
+		case int32, int64, uint32, uint64, float32, float64:
+			if v == 0 {
+				return true
+			}
+		case string:
+			if v == "" {
+				return true
+			}
+		case bool:
+			if v == false {
+				return true
+			}
+		default:
```

Comment:
I think in the default case we should return false. Otherwise we could be missing a case for some type and not realize it. For instance: I think we also need cases for `*int32`, `*string` etc?

Reply:
Having the default case return `false` (this arg is non-zero) would short-circuit the logic and prevent it from determining if later arguments were nil or zero. Having the default case return `true` (this arg is zero-ish) prevents this from being used to validate that large types (e.g. whole `*corepb.Certificates`) are non-nil. That's why I instead made sure that the list of types here is every type which can be [output by the golang proto3 code generator](https://developers.google.com/protocol-buffers/docs/proto3#scalar), so there shouldn't be any surprise types being passed in here.

I've added support for [proto2's pointer-to- types](https://developers.google.com/protocol-buffers/docs/proto#scalar), with a TODO to remove said support after the migration to proto3 is complete. Unfortunately this isn't as clean as for the non-pointer types, because of the way multi-type cases, dereferencing pointers, and type coercion interact.